testing/demo/pkg/transfer/relay.go

```diff
@@ -0,0 +1,261 @@
+package main
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	proverclient "github.com/celestiaorg/celestia-zkevm-ibc-demo/provers/client"
+	"github.com/celestiaorg/celestia-zkevm-ibc-demo/testing/demo/pkg/ethereum"
+	"github.com/celestiaorg/celestia-zkevm-ibc-demo/testing/demo/pkg/utils"
+	coretypes "github.com/cometbft/cometbft/rpc/core/types"
+	transfertypes "github.com/cosmos/ibc-go/v10/modules/apps/transfer/types"
+	"github.com/cosmos/solidity-ibc-eureka/abigen/ics26router"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// FungibleTokenPacketData represents the ICS20 token transfer data structure
+// It should match IICS20TransferMsgs.FungibleTokenPacketData in the Solidity contract
+type FungibleTokenPacketData struct {
+	Denom    string
+	Amount   *big.Int
+	Sender   string
+	Receiver string
+	Memo     string
+}
+
+// relayByTx implements the logic of an IBC relayer.
+// It processes source tx, extracts IBC events, generates proofs,
+// and creates an Ethereum transaction to submit to the ICS26Router contract.
+func relayByTx(sourceTxHash string, targetClientID string) error {
```

Comment:
ByTx you specifically mean relaying a MsgTransfer right? And this is just from Celestia to ethereum?

Reply:
Yep exactly. This is specific to a MsgTransfer in the Celestia to Ethereum direction.

I can clarify that in comments.